File: 02 medium/search problems/797_all_paths_from_source_to_target.js
/**
 * @param {number[][]} graph
 * @return {number[][]}
 */

/*

    principle: deapth first search 
    base case: node + 1 == graph.length 

*/

var allPathsSourceTarget = function(graph) {
    let paths = getPath(graph, 0, [], [])
    return paths
};

/*
    graph: the graph
    node: current node
    path: current path
    paths: all paths
 */

function getPath(graph, node, path, paths) {
    for (var j = 0; j < graph[node].length; j++) {
        getPath(graph, graph[node][j], [...path, node], paths)
    }

    if (node + 1 == graph.length) paths.push([...path, node])

    return paths;
}

File: search/node/797_all_paths_from_source_to_target.py
import copy 

class Solution(object):
    def allPathsSourceTarget(self, graph):
        """
        :type graph: List[List[int]]
        :rtype: List[List[int]]
        """
        global n, paths
        n = len(graph) 
        print(n)
        paths = []

        def dfs(node, path): 
            print(path)
            global n, paths
            if node == n - 1: 
                paths.append(path)
                return
            
            for i in graph[node]: 
                repath = copy.deepcopy(path)
                repath.append(i)
                dfs(i, repath) 
            return 
        
        for i in graph[0]: 
            dfs(i, [0, i])
        return paths
